safehttp/plugins/fetch_metadata/fetch_metadata.go

```diff
@@ -0,0 +1,155 @@
+	// Isolation Policy, but can be changed by the user.
+	//
+	// See https://web.dev/fetch-metadata/ for more information on the default policy
+	Policy func(*safehttp.IncomingRequest, map[string]bool) bool
+	// NavIsolation indicates whether the Navigation Isolation Policy should
+	// also be applied to the request as an additional layer of checking. This
+	// provides a way to mitigate clickjacking and reflected XSS by rejecting
+	// all cross-site navigations unless coming from origins where CORS is
+	// explicitly enabled. This can be used independently from a Resource
+	// Isolation Policy.
+	//
+	// WARNING: This is still an experimental feature and will be disabled by default.
+	NavIsolation bool
+	reportOnly   bool
+	logger       LoggingService
+	allowedCORS  map[string]bool
+}
+
+// NewPlugin creates a new Fetch Metadata plugin in enforce mode using the
+// defaultPolicy and a set of user-provided origins on which CORS is enabled.
+func NewPlugin(origins ...string) *Plugin {
+	m := map[string]bool{}
+	for _, origin := range origins {
+		m[origin] = true
+	}
+	return &Plugin{
+		allowedCORS: m,
+		Policy:      defaultPolicy,
+	}
+}
+
+func defaultPolicy(r *safehttp.IncomingRequest, allowedCORS map[string]bool) bool {
+	h := r.Header
+	switch h.Get("Sec-Fetch-Site") {
+	case "":
+		// Fetch Metadata is not supported by the browser so we allow the
+		// request to pass.
+		return true
+	case "same-origin", "same-site", "none":
+		// The request originated from a site served by your own server
+		// ("same-origin"), a subdomain of your site ("same-site", e.g.
+		// bar.foo.com made a request to foo.com) or was caused by the user
+		// explicitly interacting with the user-agent ("none"). Therefore it is
+		// allowed to pass.
+		return true
+	}
+
+	if m := r.Method(); h.Get("Sec-Fetch-Mode") == "navigate" && (m == safehttp.MethodGet || m == safehttp.MethodHead) {
+		if dest := h.Get("Sec-Fetch-Dest"); dest == "object" || dest == "embed" {
+			// The request is cross-site and originates from <object> or <embed>
+			// so it is rejected.
+			return false
+		}
+		// The request is cross-site, but a simple top-level navigation so we
+		// allow it to pass.
+		return true
+	}
+
+	origin := h.Get("Origin")
+	if _, ok := allowedCORS[origin]; ok && h.Get("Sec-Fetch-Mode") == "cors" {
+		// The request is cross-site but sent from an
+		// origin where Cross-Origin Resource Sharing is explicitly allowed.
+		return true
+	}
+
+	// The request is cross-site, not navigational and sent from an endpoint
+	// where CORS is not enabled therefore it is rejected.
+	return false
+}
+
+func navigationIsolationPolicy(r *safehttp.IncomingRequest, allowedCORS map[string]bool) bool {
+	h := r.Header
+	origin := h.Get("Origin")
+	if h.Get("Sec-Fetch-Site") == "cross-site" && h.Get("Sec-Fetch-Mode") == "navigate" {
+		if _, ok := allowedCORS[origin]; !ok {
+			return false
+		}
+	}
+	return true
```

Comment:
The cors-allowed check should be the first one.

With regards to the policy you expressed, I don't think this is what we should go for.

Since this policy has to run on top of the RI one I would say this should be something like:
```go
// Omitted: check if CORS allowed

if site == "cross-site" && navigationalMode[mode] {
  return false
}
return resourceIsolationPolicy(r, allowedCORS)
```